README.md
jQuery Easy Overlay Plugin
=====================

Easy & Simple jQuery plugin for overlay an existing element with Css & FontAwesome

----------

Requirements
---------
	- JQuery
	- FontAwesome

Getting Started
---------

Include the script on your page

```html
<script src="js/jquery.easy-overlay.js">
```

And, for overlay an existing element
```javascript
$("#elementDesired").easyOverlay("start");
```

For remove the overlay layer
```javascript
$("#elementDesired").easyOverlay("stop");
Or
$.easyOverlay("stop");
```

That's all!

Options
---------

To customize some aspects of the plugin behaviour, you can use options. Defaults are:
```
{
	zindex: "99999",
	spin: true,
	delay: 400
}
```

You can pass an object as a second parameter
```
$("#elementDesired").easyOverlay("start", { /* Your options */ });
```

And/or set your default values, for all calls of easyOverlay, without passing arguments:
```
jQuery.fn.easyOverlay.options = { /* Here your options */ };
```

Options explained:
```
    zindex: 99999, //The default z-index of the overlay is 99999
    spin:
        true // Show the fontawesome loading spinner, it's the default behaviour
        false // No spinner will be shown
        { url: 'image-web-url', width: 'image-width', height: 'image-height' } // A custom spinner, with an image from a url; sizes can be specified with units or percentages
    delay: 200     //Time in miliseconds for the fadeIn or fadeOut effects of the overlay.
});
```

And set your default values, for all calls of easyOverlay, without passing arguments:
```javascript
jQuery.fn.easyOverlay.options = { /* Here your options */ };
```

**Note**:
  * When using the `stop` option, actually only the `delay` parameter will be relevant;
  * Options passed as arguments will override those setted within jQuery.fn.easyOverlay.options property;



CHANGELOG.md
# CHANGELOG

## 1.0.0 (Aug 1, 2015)
* Support two methods: overlay() and overlayout()

## 2.0.0 (Aug 1, 2015)
* Only one method: easyOverlay()
* 1 option 'action' with values:
  * 'start'
  * 'stop'

## 2.1.0 (Oct 26, 2016)
* Options introducted, to extend the plugin behaviour

## 2.2.0 (Oct 26, 2016)
* The overlay now works also on multiple DOM elements

## 2.3.0 (Nov 3, 2016)
* Is possible specify a custom image as spinner, along with its dimensions (with units or percentages)

  ### 2.3.1 (Nov 7, 2016)
  * ZIndex on the container was not set, bug fixed

  ### 2.3.2 (Nov 19, 2016)
  * Now, default parameters can be set within jQuery.fn.easyOverlay.options

  ### 2.3.3 (Dec 17, 2016)
  * Perfect centering of the spinner container


old_versions/jquery.easy-overlay.1.0.0.js
(function( $ ){
	$.fn.overlay = function() {
    	var $t = $(this);
		
		if($t.length <= 0) return;
		
	  	var vzIndex;
	  	if($t.css('z-index') == "auto") vzIndex = "99999";
	  	else vzIndex = parseFloat($t.css('z-index')) + 1;
		
		if( !$("#jqueryEasyOverlayDiv").length ) {
			var innerDiv = document.createElement('div');
			$(innerDiv).css({ position: "absolute"}).attr("id", "jqueryOverlayLoad").html("<i class='fa fa-spin fa-spinner fa-2x'></i>&nbsp;");
			
			var div = document.createElement('div');	
			$(div).css({
				display: "none",
				position: "absolute",
				background: "#fff"
			}).attr('id',"jqueryEasyOverlayDiv").append(innerDiv);
			
			$("body").append(div);
		}
		
		
		
	  	$("#jqueryEasyOverlayDiv").css({
			opacity : 0.5,
		  	zIndex  : vzIndex,
		  	top     :  $t.offset().top,
		  	left    : $t.offset().left,
		  	width   : $t.outerWidth(),
		  	height  : $t.outerHeight()
		});
		
		var topOverlay = (($t.height()/2)-12);
		if(topOverlay < 0) topOverlay = 0;
		$("#jqueryOverlayLoad").css({
			top  : topOverlay,
		  	left : (($t.width()/2)-12)
		});
		
		$("#jqueryEasyOverlayDiv").fadeIn();

		return this;
   	}; 
})(jQuery);

(function( $ ){
	$.fn.overlayout = function() {
		if( $("#jqueryEasyOverlayDiv").length ) $("#jqueryEasyOverlayDiv").fadeOut();
		return this;
   }; 
})(jQuery);


jquery.easy-overlay.js
/*
JQuery Easy Overlay v2.3.3

Copyright 2015 Nazzareno Mazzitello, Andrés Cabrera and Contributors

This file is part of jQuery Easy Overlay.

jQuery Easy Overlay is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

jQuery Easy Overlay is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with jQuery Easy Overlay.  If not, see <http://www.gnu.org/licenses/>.
*/
(function( $ ){
	$.fn.easyOverlay = function(action, options) {

		// Preparing options
		var defaults = {
            zindex: "99999",
            spin: true,
            delay: 400,
        };
        var _options = $.extend(true, defaults, $.fn.easyOverlay.options || {}, options || {});

		// States constants
		var STATE_VISIBLE = 1;
		var STATE_HIDDEN = 0;
		
		function init(target) {
			if(target.length <= 0 || target.data('easyOverlayState') == STATE_VISIBLE) return;
			target.data('easyOverlayInitiated', true);
			$.fn.easyOverlay.indexCounter++;
			var easyOverlayIndex = $.fn.easyOverlay.indexCounter;
			
			// Calculating OVERLAY DIV z-index
			var overlayZIndex;
			var targetZIndex = target.css('z-index');
			if (targetZIndex == "auto")
				overlayZIndex = _options.zindex;
			else 
				overlayZIndex = parseFloat(targetZIndex) + 1;
			
			// Putting and Styling OVERLAY DIV if doesn't exist
			if( !$("#jqueryEasyOverlayDiv"+easyOverlayIndex).length ) {
				var innerDiv = document.createElement('div');
				if (typeof _options.spin == 'object') {
					$(innerDiv)
					.css({ position: "absolute", "background-image": "url(" + _options.spin.url + ")", "background-size": "100%", width: _options.spin.width, height: _options.spin.height })
					.attr("id", "jqueryOverlayLoad"+easyOverlayIndex);
				} else if (_options.spin) {
					$(innerDiv)
					.css({ position: "absolute" })
					.attr("id", "jqueryOverlayLoad"+easyOverlayIndex)
					.html("<i class='fa fa-spin fa-spinner fa-2x'></i>&nbsp;");
				} else {
					$(innerDiv)
					.css({ position: "absolute" })
					.attr("id", "jqueryOverlayLoad"+easyOverlayIndex);
				}
				
				
				var containerDiv = document.createElement('div');	
				$(containerDiv)
					.css({
						display: "none",
						position: "absolute",
						background: "#fff"
					})
					.attr('id',"jqueryEasyOverlayDiv"+easyOverlayIndex)
					.append(innerDiv);
				
				$("body").append(containerDiv);
			}

			// Calculating Spinner Div positioning
			var spinWidth = _options.spin.width.slice(0, _options.spin.width.length - 2);
			var spinHeight = _options.spin.height.slice(0, _options.spin.height.length - 2);
			var spinTop = ((target.height()/2)-(spinWidth / 2));
			var spinLeft = ((target.width()/2)-(spinHeight / 2));
			if(spinTop < 0) spinTop = 0;
			$("#jqueryOverlayLoad"+easyOverlayIndex).css({
				top  : spinTop,
				left : spinLeft
			});
			target.data('easyOverlayIndex', easyOverlayIndex);
		}
		
		function start(target) {
			var easyOverlayIndex = target.data('easyOverlayIndex');
			// Restoring some CSS of OVERLAY DIV after every 'stop' because jquery.fadeOut method take off it
			$("#jqueryEasyOverlayDiv"+easyOverlayIndex).css({
				opacity : 0.5,
				top     : target.offset().top,
				left    : target.offset().left,
				width   : target.outerWidth(),
				height  : target.outerHeight()
			});
			
			// Show OVERLAY DIV
			$("#jqueryEasyOverlayDiv"+easyOverlayIndex).fadeIn(_options.delay);
			target.data('easyOverlayState', STATE_VISIBLE);
		}

		function stop(target) {
			var easyOverlayIndex = target.data('easyOverlayIndex');
			if( $("#jqueryEasyOverlayDiv"+easyOverlayIndex).length ) {
				$("#jqueryEasyOverlayDiv"+easyOverlayIndex).fadeOut(_options.delay);
			}
			target.data('easyOverlayState', STATE_HIDDEN);
		}

		switch (action) {
			case 'start':
				{
					if(this.length == 1) {
						if (typeof this.data('easyOverlayInitiated') == 'undefined') {
							init(this);
						}
						start(this);
					}
				}
				break;
			case 'stop':
				{
					if (this.length == 1) {
						stop(this);
					}
				}
				break;
		}
    	
		return this;
			
	}
	$.fn.easyOverlay.indexCounter = 0;
})(jQuery);
